### PR TITLE
Add benchmark framework to `sway-lsp`

### PR DESCRIPTION
**Pull request by JoshuaBatty**
## Description
This PR adds a way to benchmark the LSP requests and `TokenMap` methods using `criterion`. I've also added a large sway benchmark project that can be used when running benchmarks now and in the future to measure performance improvements. 

Here is the result of the benchmarks. 

| Request Type         | duration  |
|----------------------|---------------|
| hover                | 442ms    |
documentHighlight    | 441ms   |
| definition           | 433ms   |
 prepareRename        | 433ms   |
| rename               | 436ms    |
 semantic_tokens        | 464ms   |
| document_symbol               | 447ms    |
 completion        | 893ms   |
| inlay_hints               | 15ms    |
 code_action        | 444ms   |
| code_lens        | 34ms   |
 on_enter               | 572ms    |

| TokenMap method         | duration  |
|----------------------|---------------|
| tokens_for_file                | ~~230ns~~ 10ms    |
idents_at_position    | 438ms   |
| tokens_at_position           | 437ms   |
 token_at_position        | 439ms   |
| parent_decl_at_position               | 437ms    |

| Compilation        | duration  |
|----------------------|---------------|
| compile_and_traverse                | 1.795sec    |

related to #4901 

## Checklist

- [x] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [ ] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by anton-trunov** [comment]
LGTM! Perhaps we should add some kind of developer documentation where it says we actually have benchmarks and how to run those?

File: sway-lsp/src/core/session.rs [hunk 1]
```diff
@@ -86,6 +86,12 @@ pub struct Session {
     pub diagnostics: Arc<RwLock<Diagnostics>>,
 }
 
+impl Default for Session {
```
Comment (xunilrj) [inline sway-lsp/src/core/session.rs:89]: Do we need both `default()` and `new()`?

**Comment by xunilrj** [comment]
Some methods like `tokens_for_file` are lazy. Maybe that is why their result is so good. You should `collect()` its results to a `Vec`.

I am also wondering if it is worth to serialize responses to `JSON`. Because this way we also assess the "size" of the response.

File: sway-lsp/benches/lsp_benchmarks/requests.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    c.bench_function("goto_definition", |b| {
+        b.iter(|| session.token_definition_response(uri.clone(), position))
+    });
+
+    c.bench_function("prepare_rename", |b| {
+        b.iter(|| capabilities::rename::prepare_rename(session.clone(), uri.clone(), position))
+    });
+
+    c.bench_function("rename", |b| {
```
Comment (sdankel) [inline sway-lsp/benches/lsp_benchmarks/requests.rs:28]: Should we do inlay hints as well, since it's highly visible to users?

**Review by sdankel — APPROVED** [review]
Awesome work! 

If the benchmarks don't take too long to run, it would be cool to integrate the request benchmarks into CI with a script like [this](https://gist.github.com/yamafaktory/a3512b379c8caf7fd999ac7c25d0a1df#file-after-success-sh). We could detect any change in the compiler or LSP code that significantly slows things down.

File: sway-lsp/src/core/session.rs (hunk 1 above)
Comment (JoshuaBatty) [reply sway-lsp/src/core/session.rs:89]: I only added this to appease clippy. Seems like this lint has recently been added.

**Comment by JoshuaBatty** [comment]
> Some methods like `tokens_for_file` are lazy. Maybe that is why their result is so good. You should `collect()` its results to a `Vec`.

Good point, will modify the benchmark to collect into a Vec for this one. 
I'll have a think about measuring the size and address that in a later PR. Thanks for the review!

File: sway-lsp/benches/lsp_benchmarks/requests.rs (hunk 2 above)
Comment (JoshuaBatty) [reply sway-lsp/benches/lsp_benchmarks/requests.rs:28]: Yeah good idea, i'll add inlay hints to benchmark also.

Comment (JoshuaBatty) [reply sway-lsp/benches/lsp_benchmarks/requests.rs:28]: I've just added in the missing requests and updated the results

**Comment by JoshuaBatty** [comment]
> If the benchmarks don't take too long to run, it would be cool to integrate the request benchmarks into CI with a script like [this](https://gist.github.com/yamafaktory/a3512b379c8caf7fd999ac7c25d0a1df#file-after-success-sh). We could detect any change in the compiler or LSP code that significantly slows things down.

Nice, yeah something like that could be cool to integrate into our CI down the track. Will have a think about what would be useful for us.

**Comment by JoshuaBatty** [comment]
> LGTM! Perhaps we should add some kind of developer documentation where it says we actually have benchmarks and how to run those?

I've added a page under the Engineering/LSP page in the internal Notion docs for now. We could perhaps add another .MD to this repo at some point if the need arrises. 

If you want to run them yourself just `cd` into the `sway-lsp` directory and run `cargo bench`

**Review by sdankel — APPROVED** [review]
:shipit:

**Review by sdankel — APPROVED** [review]

**Review by xunilrj — APPROVED** [review]

**Pull request merged**